database/engine/metadata_log/metadatalogapi.c

```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+#include "metadatalog.h"
+
+/* The buffer must not be empty */
+static void metalog_commit_record(struct metalog_instance *ctx, BUFFER *buffer, enum metalog_opcode opcode)
+{
+    struct metalog_cmd cmd;
+
+    assert(buffer_strlen(buffer));
+    assert(opcode == METALOG_COMMIT_CREATION_RECORD || opcode == METALOG_COMMIT_DELETION_RECORD);
+
+    cmd.opcode = opcode;
+    cmd.record_io_descr.buffer = buffer;
+    metalog_enq_cmd(&ctx->worker_config, &cmd);
+}
+
+static inline void metalog_commit_creation_record(struct metalog_instance *ctx, BUFFER *buffer)
+{
+    metalog_commit_record(ctx, buffer, METALOG_COMMIT_CREATION_RECORD);
+}
+
+static inline void metalog_commit_deletion_record(struct metalog_instance *ctx, BUFFER *buffer)
+{
+    metalog_commit_record(ctx, buffer, METALOG_COMMIT_DELETION_RECORD);
+}
+
+void metalog_commit_update_host(RRDHOST *host)
+{
+    struct metalog_instance *ctx;
+    BUFFER *buffer;
+
+    /* Metadata are only available with dbengine */
+    if (!host->rrdeng_ctx)
```

Comment:
The lock in L39 below guards against the read accesses to the host for the sprintf but not this or L36.

Reply:
This code has been taken from streaming. The host lock is primarily protects the list of charts. Moreover, it has been use to protect the list of labels when labels were added and in some cases to protect some writeable fields in the host structure.

However, most of those things are not 100% safe as we proved when we crashed the agent. The immutable values however are never protected in the current locking scheme. One thing that never changes is exactly what is in L36 which is `host->rrdeng_ctx->metalog_ctx`.

This is a way too big discussion to have but I would argue that protecting L36 and many more like it does not give any protection with the way things are currently.